pkg/logql/sharding.go

```diff
@@ -55,3 +62,255 @@ func (c ConcatLogSelectorExpr) String() string {
 
 	return fmt.Sprintf("%s ++ %s", c.LogSelectorExpr.String(), c.next.String())
 }
+
+type Shards []astmapper.ShardAnnotation
+
+func (xs Shards) Encode() (encoded []string) {
+	for _, shard := range xs {
+		encoded = append(encoded, shard.String())
+	}
+
+	return encoded
+}
+
+// ParseShards parses a list of string encoded shards
+func ParseShards(strs []string) (Shards, error) {
+	if len(strs) == 0 {
+		return nil, nil
+	}
+	shards := make([]astmapper.ShardAnnotation, 0, len(strs))
+
+	for _, str := range strs {
+		shard, err := astmapper.ParseShard(str)
+		if err != nil {
+			return nil, err
+		}
+		shards = append(shards, shard)
+	}
+	return shards, nil
+}
+
+// Downstreamer is an interface for deferring responsibility for query execution.
+// It is decoupled from but consumed by a downStreamEvaluator to dispatch ASTs.
+type Downstreamer interface {
+	Downstream(Expr, Params, Shards) (Query, error)
+}
+
+// DownstreamEvaluator is an evaluator which handles shard aware AST nodes
+type DownstreamEvaluator struct {
+	Downstreamer
+	defaultEvaluator *DefaultEvaluator
+}
+
+func NewDownstreamEvaluator(downstreamer Downstreamer) *DownstreamEvaluator {
+	return &DownstreamEvaluator{
+		Downstreamer: downstreamer,
+		defaultEvaluator: NewDefaultEvaluator(
+			QuerierFunc(func(_ context.Context, p SelectParams) (iter.EntryIterator, error) {
+				// TODO(owen-d): add metric here, this should never happen.
```

Comment:
I dont' think you want to add a metrics, can't you use a nil Querier if it's not used ? or a NoopQuerier ?